src/ObjectUtils/CoffFile.cpp

```diff
@@ -0,0 +1,150 @@
+// found in the LICENSE file.
+
+#include "ObjectUtils/CoffFile.h"
+
+#include <absl/strings/str_format.h>
+#include <llvm/Demangle/Demangle.h>
+#include <llvm/Object/Binary.h>
+#include <llvm/Object/COFF.h>
+#include <llvm/Object/ObjectFile.h>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Result.h"
+#include "symbol.pb.h"
+
+namespace orbit_object_utils {
+
+namespace {
+
+using orbit_grpc_protos::ModuleSymbols;
+using orbit_grpc_protos::SymbolInfo;
+
+class CoffFileImpl : public CoffFile {
+ public:
+  CoffFileImpl(std::filesystem::path file_path,
+               llvm::object::OwningBinary<llvm::object::ObjectFile>&& owning_binary);
+  [[nodiscard]] ErrorMessageOr<orbit_grpc_protos::ModuleSymbols> LoadDebugSymbols() override;
+  [[nodiscard]] bool HasDebugSymbols() const override;
+  [[nodiscard]] std::string GetName() const override;
+  [[nodiscard]] std::filesystem::path GetFilePath() const override;
+
+ private:
+  ErrorMessageOr<SymbolInfo> CreateSymbolInfo(const llvm::object::SymbolRef& symbol_ref);
+  const std::filesystem::path file_path_;
+  llvm::object::OwningBinary<llvm::object::ObjectFile> owning_binary_;
+  llvm::object::COFFObjectFile* object_file_;
+  bool has_symbol_table_;
+};
+
+CoffFileImpl::CoffFileImpl(std::filesystem::path file_path,
+                           llvm::object::OwningBinary<llvm::object::ObjectFile>&& owning_binary)
+    : file_path_(std::move(file_path)),
+      owning_binary_(std::move(owning_binary)),
+      has_symbol_table_(false) {
+  object_file_ = llvm::dyn_cast<llvm::object::COFFObjectFile>(owning_binary_.getBinary());
+  // Per specification, a COFF file has a symbol table if the pointer to the symbol table is not 0.
+  has_symbol_table_ = (object_file_->getSymbolTable() != 0);
+}
+
+ErrorMessageOr<SymbolInfo> CoffFileImpl::CreateSymbolInfo(
+    const llvm::object::SymbolRef& symbol_ref) {
+  if ((symbol_ref.getFlags() & llvm::object::BasicSymbolRef::SF_Undefined) != 0) {
+    return ErrorMessage("Symbol is defined in another object file (SF_Undefined flag is set).");
+  }
+  const std::string name = symbol_ref.getName() ? symbol_ref.getName().get() : "";
+  // Unknown type - skip and generate a warning.
+  if (!symbol_ref.getType()) {
+    LOG("WARNING: Type is not set for symbol \"%s\" in \"%s\", skipping.", name,
+        file_path_.string());
+    return ErrorMessage(absl::StrFormat(R"(Type is not set for symbol "%s" in "%s", skipping.)",
+                                        name, file_path_.string()));
+  }
+  // Limit list of symbols to functions. Ignore sections and variables.
+  if (symbol_ref.getType().get() != llvm::object::SymbolRef::ST_Function) {
+    return ErrorMessage("Symbol is not a function.");
+  }
+
+  // Symbol addresses are relative virtual addresses (RVAs) that are relative to the
+  // start of the section. To get the address relative to the start of the object
+  // file, we need to find the section and add the base address of the section.
```

Comment:
extract to a separate function? Also it is kind of counterintuitive: are there multiple sections with symbols? I was under impression there is one.

Reply:
I don't know if it happens in practice, but the specifications mentions multiple text sections. 

For example, this text part mentions multiple sections: 
"SizeOfCode Size of the code (text) section, or the sum of all code sections if there are multiple sections." (**https://courses.cs.washington.edu/courses/cse378/03wi/lectures/LinkerFiles/coff.pdf**)

Looking up the section is also not expensive, it's all just indices returned and some index arithmetic (it's not iterating over the sections to find the section, for example).

I will pull this out into a separate function, though.

Reply:
Done.